src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs

```diff
@@ -0,0 +1,2087 @@
+            SyntaxNode syntax = boundReturnStatement.Syntax;
+            ITypeSymbol type = null;
+            Optional<object> constantValue = default(Optional<object>);
+            return new LazyReturnStatement(returnedValue, isInvalid, syntax, type, constantValue);
+        }
+        private static IReturnStatement CreateBoundYieldReturnStatementOperation(BoundYieldReturnStatement boundYieldReturnStatement)
+        {
+            Lazy<IOperation> returnedValue = new Lazy<IOperation>(() => (IOperation)Create(boundYieldReturnStatement.Expression));
+            bool isInvalid = boundYieldReturnStatement.HasErrors;
+            SyntaxNode syntax = boundYieldReturnStatement.Syntax;
+            ITypeSymbol type = null;
+            Optional<object> constantValue = default(Optional<object>);
+            return new LazyReturnStatement(returnedValue, isInvalid, syntax, type, constantValue);
+        }
+        private static ILockStatement CreateBoundLockStatementOperation(BoundLockStatement boundLockStatement)
+        {
+            Lazy<IOperation> lockedObject = new Lazy<IOperation>(() => (IOperation)Create(boundLockStatement.Argument));
+            Lazy<IOperation> body = new Lazy<IOperation>(() => (IOperation)Create(boundLockStatement.Body));
+            bool isInvalid = boundLockStatement.HasErrors;
+            SyntaxNode syntax = boundLockStatement.Syntax;
+            ITypeSymbol type = null;
+            Optional<object> constantValue = default(Optional<object>);
+            return new LazyLockStatement(lockedObject, body, isInvalid, syntax, type, constantValue);
+        }
+        private static IInvalidStatement CreateBoundBadStatementOperation(BoundBadStatement boundBadStatement)
+        {
+            Lazy<ImmutableArray<IOperation>> children = new Lazy<ImmutableArray<IOperation>>(() => boundBadStatement.ChildBoundNodes.SelectAsArray(n => Create(n)));
+            bool isInvalid = boundBadStatement.HasErrors;
+            SyntaxNode syntax = boundBadStatement.Syntax;
+            ITypeSymbol type = null;
+            Optional<object> constantValue = default(Optional<object>);
+            return new LazyInvalidStatement(children, isInvalid, syntax, type, constantValue);
+        }
+        private static IVariableDeclarationStatement CreateBoundLocalDeclarationOperation(BoundLocalDeclaration boundLocalDeclaration)
+        {
+            Lazy<ImmutableArray<IVariableDeclaration>> declarations = new Lazy<ImmutableArray<IVariableDeclaration>>(() => GetVariableDeclarationStatementVariables(boundLocalDeclaration));
+            bool isInvalid = boundLocalDeclaration.HasErrors;
+            SyntaxNode syntax = boundLocalDeclaration.Syntax;
+            ITypeSymbol type = null;
+            Optional<object> constantValue = default(Optional<object>);
+            return new LazyVariableDeclarationStatement(declarations, isInvalid, syntax, type, constantValue);
+        }
+        private static IVariableDeclarationStatement C
```

Comment:
@heejaechang You may be right that LINQ may be cheaper than the combination of `SelectAs` and `WhereAs`. For LINQ, I suspect there are three allocations: the `Select` result, `Where` result, and `ToImmutableArray()` result. The alternative should be at most two allocations: one for `SelectAs`, and zero or one for `WhereAs` depending on whether `WhereAs` a non-empty subset, That said, if there are two allocations, the total bytes allocated may be larger than the LINQ approach. Feel free to keep as is.